pyccel/ast/basic.py

```diff
@@ -165,7 +165,90 @@ def get_attribute_nodes(self, search_type, excluded_nodes = ()):
         self._recursion_in_progress = False
         return results
 
-    def substitute(self, original, replacement, excluded_nodes = ()):
+    def search_user_nodes(self, search_node, excluded_nodes = ()):
+        """ Identifies whether this object exists within the object search_node
+
+        Parameters
+        ----------
+        search_node : Basic
+                      The object which we are looking for
+        excluded_nodes : tuple of types
+                      Types for which get_user_nodes should not be called
```

Comment:
This makes sense, and I agree that a better name is needed. However do you think `is_attribute_of` is clear enough about the recursive nature of this function? I.e. for a node:
`Assign(Variable('a'), PyccelAdd(Variable('b'), Variable('c'))`
do you expect `var_b.is_attribute_of(assign)` to return `True` (as would be the case with the function here)?

Reply:
I see that in your example `var_b` is not a "direct" attribute of the `assign` node, but rather an attribute of an attribute...

At any rate, given that most of the operations on user/attribute nodes are recursive by nature, I don't see a reason to worry (provided that we document the functions properly).

In order be completely rigorous and clear, one might want to add the keyword-only argument `recursive` to this function. But since we would always pass the `True` value I see no reason to make this more complicated than it already is.

Reply:
Fixed 21f8afa9

Reply:
By the way, which option do you think is faster?

1. Recursively search the user nodes of `self` for `search_node`
2. Recursively check if `self` is an attribute node of `search_node`

Reply:
Searching through the user nodes should always be faster as most objects have 1 user node, but many attribute nodes. I didn't consider this when I wrote the code. I wonder if I should modify `is_user_node` to be:
```
def is_user_node(self, search_node):
    return search_node.is_attribute_node(self)
```

Reply:
So I think that we should actually change `is_user_of` to search for `self` through the user nodes of `search_node` (we are currently searching for `search_node` through the attribute nodes of `self` instead), and then modify `is_attribute_of`:
```python
def is_attribute_of(self, node):
    return node.is_user_of(self)
```

Reply:
Fixed a8662ed38ad2ee9e268966efc649a1fb8b137b03

Reply:
Thanks! But apparently Codacy was not happy of my suggestion 😅

Reply:
Fixed 018badbcd768f514e804131ac49219c6a9613680

Reply:
Thanks. Please update the docstrings!

Reply:
Fixed 6d78abaa